### PR TITLE
Use access control to filter out redirected tables

### PR DESCRIPTION
**Pull request by lukasz-walkiewicz**
## Description
Consult access control when getting comments for redirected tables.

Fixes https://github.com/trinodb/trino/issues/18514

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Security
* Added access control check to redirected tables when getting comments. ({issue}`#18514`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java [hunk 1]
```diff
@@ -122,19 +124,19 @@ public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, Connect
                         prefix.getSchemaName(),
                         relationNames -> accessControl.filterTables(session.toSecurityContext(), catalog, relationNames));
 
+                Map<SchemaTableName, TableMetadata> tables = getRedirectionAwareTables(session, prefix);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java:127]: Is this listing all tables matching prefix the second time?

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java:127]: indeed `io.trino.plugin.iceberg.catalog.glue.TestIcebergGlueCatalogAccessOperations#testSystemMetadataTableComments` regresses. 

not only listing tables is redundant; getRedirectionAwareTables acquires a table handle for every each of them, and getting table handle is costly. This incurs additional GetTable call (avoidable when number of table is small and we can cache all them in per-query cache... which we should assume to be unbounded) and additional S3 trips.

**Review by findepi — CHANGES_REQUESTED** [review]

**Comment by findepi** [comment]
BTW redirects are implemented rather inefficiently today. 

1. We detect a redirect, so we know where it leads to, and record only the fact there is a redirect so we ask for same information again.
2. We don't try to do any bulk retrievals in case of redirects. Thus if i use hive/iceberg mixed metastore with fifty-fifty split, 50% info can be retrieved efficiently (if right APIs implemented), but remaining 50% is going inefficient code path.

Then the original problem https://github.com/trinodb/trino/issues/18514 seems to apply to redirects only, so I'd try to isolate the fix to change the code path only for the redirects. It sounds defensive, I know, but the code path for standard case is pretty optimized at this point. 
(BTW can someone confirm my expected behavior in that issue is right? Or maybe the current behavior is right??)

**Comment by lukasz-walkiewicz** [comment]
Thanks for the review. Those are great comments, I was, rather unnecessary, optimizing for the least amount of calls to `filterTables` whereas it's obviously, as you rightly pointed out, not the most expensive operation here.
I've tried a different approach, please take another look.

File: core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java [hunk 2]
```diff
@@ -159,13 +157,27 @@ private RelationComment getRelationComment(Session session, QualifiedObjectName
             return new RelationComment(true, view.get().getComment());
         }
 
-        RedirectionAwareTableHandle redirectionAware = metadata.getRedirectionAwareTableHandle(session, relationName);
-        if (redirectionAware.tableHandle().isPresent()) {
-            // TODO (https://github.com/trinodb/trino/issues/18514) this should consult accessControl on redirected name. Leaving for now as-is.
-            return new RelationComment(true, metadata.getTableMetadata(session, redirectionAware.tableHandle().get()).getMetadata().getComment());
+        return getCommentOnRedirectedTable(session, relationName);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java:160]: "getTableCommentRedirectionAware" or "getTableComment"

(at this point we don't know whether table is redirected)

File: core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java [hunk 3]
```diff
@@ -159,13 +157,27 @@ private RelationComment getRelationComment(Session session, QualifiedObjectName
...
+    }
+
+    private RelationComment getCommentOnRedirectedTable(Session session, QualifiedObjectName relationName)
+    {
+        String catalog = relationName.getCatalogName();
+        SchemaTableName name = relationName.asSchemaTableName();
+        RedirectionAwareTableHandle redirectedTable = metadata.getRedirectionAwareTableHandle(
+                session,
+                new QualifiedObjectName(catalog, name.getSchemaName(), name.getTableName()));
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java:169]: is it the same as `relationName` argument?

File: core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java [hunk 4]
```diff
@@ -159,13 +157,27 @@ private RelationComment getRelationComment(Session session, QualifiedObjectName
...
+
+        if (redirectedTable.redirectedTableName().isEmpty() || redirectedTable.tableHandle().isEmpty()) {
+            return new RelationComment(false, Optional.empty());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java:172]: we end up existing here when `prefix.getTableName().isPresent()` and table is not redirected.
doesn't look correct.

File: core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java [hunk 5]
```diff
@@ -159,13 +157,27 @@ private RelationComment getRelationComment(Session session, QualifiedObjectName
...
+        }
+
+        SchemaTableName redirectedTableName = redirectedTable.redirectedTableName().get().asSchemaTableName();
+        if (!accessControl.filterTables(session.toSecurityContext(), catalog, ImmutableSet.of(redirectedTableName)).contains(redirectedTableName)) {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java:176]: the redirected table is (typically) in a different catalog, so `catalog` doesn't seem right here

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 6]
```diff
@@ -583,6 +613,15 @@ public void testCommentView()
         assertAccessAllowed("COMMENT ON VIEW " + viewName + " IS 'new comment'");
     }
 
+    @Test
+    public void testCommentOnRedirectedTable()
+    {
+        @Language("SQL") String query = "SELECT * FROM system.metadata.table_comments WHERE catalog_name = 'mock' AND schema_name = 'default' AND table_name LIKE 'redirected%'";
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:619]: `@Language("SQL")` is redundant to get SQL syntax highliting in intelliJ (IDE infers this based on the usage)

File: core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java (hunk 4 above)
Comment (lukasz-walkiewicz) [reply core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java:173]: spot on! thanks

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 7]
```diff
@@ -212,6 +224,23 @@ public Map<SchemaTableName, ConnectorMaterializedViewDefinition> apply(Connector
...
+                    }
+                    else {
+                        return Optional.empty();
+                    }
+                }))
+                .withGetComment((schemaTableName -> {
+                    if (schemaTableName.getTableName().equals(REDIRECTED_TARGET)) {
+                        return Optional.of("this is a redirected table");
+                    }
+                    else {
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:240]: redundant `else`

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 8]
```diff
@@ -212,6 +224,23 @@ public Map<SchemaTableName, ConnectorMaterializedViewDefinition> apply(Connector
                 .withColumnProperties(() -> ImmutableList.of(
                         integerProperty("another_property", "description", 0, false),
                         stringProperty("string_column_property", "description", "", false)))
+                .withRedirectTable(((connectorSession, schemaTableName) -> {
+                    if (schemaTableName.equals(SchemaTableName.schemaTableName(DEFAULT_SCHEMA, REDIRECTED_SOURCE))) {
+                        return Optional.of(
+                                new CatalogSchemaTableName("mock", SchemaTableName.schemaTableName(DEFAULT_SCHEMA, REDIRECTED_TARGET)));
+                    }
+                    else {
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:232]: redundant `else`

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 9]
```diff
@@ -157,6 +160,15 @@ protected SystemAccessControl delegate()
                     }
                     return new MockConnectorTableHandle(schemaTableName);
                 })
+                .withListSchemaNames((connectorSession -> ImmutableList.of(DEFAULT_SCHEMA)))
+                .withListTables((connectorSession, schemaName) -> {
+                    if (schemaName.equals(DEFAULT_SCHEMA)) {
+                        return ImmutableList.of(REDIRECTED_SOURCE);
+                    }
+                    else {
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:168]: redundant `else`

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
/test-with-secrets sha=b9e885c2bffe09c74f0f4e9d3fcb8bd5d65856fb

**Comment by lukasz-walkiewicz** [comment]
Fixed, thanks!

**Comment by findepi** [comment]
Thanks!

**Pull request merged**